python_modules/libraries/dagster-gcp/dagster_gcp/bigquery/resources.py

```diff
@@ -96,3 +101,39 @@ def bigquery_resource(context):
     bq_resource = BigQueryResource.from_resource_context(context)
     with bq_resource.get_client() as client:
         yield client
+
+
+def get_data_freshness(
+    context: OpExecutionContext,
+    client: bigquery.Client,
+    dataset_id: str,
+    table_ids: Optional[Sequence[str]] = None,
+) -> dict[str, Optional[float]]:
+    """Get the freshness of the tables in a dataset.
+
+    Args:
+        client: BigQuery client.
+        dataset_id: Dataset ID.
+        table_ids: List of table IDs. If not provided, all tables in the dataset will be checked.
+
+    Returns:
+        Dictionary of table IDs to the freshness of the tables.
+    """
+    context.log.info(f"Getting data freshness for dataset {dataset_id}")
+    try:
+        dataset = client.get_dataset(dataset_id)
+    except:
+        context.log.error(f"Dataset {dataset_id} not retrieved")
+        return {}
+
+    all_table_ids = [table.table_id for table in client.list_tables(dataset)]
```

Comment:
when a list of table IDs is provided, it seems potentially wasteful to fetch metadata about tables that weren't provided by the user. are we able to avoid that?

Reply:
I think this comment is no longer relevant given I have switched to using a single table per call of this function.